Limit orders/receipt

nenadV91 opened the pull request:
# Summary

Implements #1338

This PR adds implementation of receipt modal

# To test
- go to limit-orders page
- create a new order or open order history tab
- click on some order in the table
- the new receipt modal should open

![Screenshot from 2022-11-24 18-12-49](https://user-images.githubusercontent.com/34926005/203838089-c4f62c75-1b65-45bc-bfd2-098d3a42d437.png)

shoom3301 commented on src/cow-react/api/gnosisProtocol/api.ts:
```diff
@@ -608,6 +609,7 @@ function transformEthFlowOrder(order: OrderMetaData): OrderMetaData {
 
 // Register some globals for convenience
 registerOnWindow({
+  JSBI,
```
Looks like redundant

nenadV91 replied:
Yep, good catch

shoom3301 commented on src/custom/hooks/useActivityDerivedState.ts:
```diff
@@ -79,7 +79,7 @@ function getActivityDerivedState(props: {
 }
 
-function getActivityLinkUrl(params: {
+export function getActivityLinkUrl(params: {
```
If it's not an override of UNI-flies, let's move it to cow-react

nenadV91 replied:
I think we should do that in another PR, this one is large enough.

shoom3301 commented on src/cow-react/modules/limitOrders/containers/OrdersWidget/index.tsx:
```diff
@@ -59,11 +60,14 @@ export function OrdersWidget() {
+        balancesAndAllowances={pendingBalancesAndAllowances}
+        isWalletConnected={!!account}
+      />
+      <LimitOrdersReceiptModal />
```
Would be awesome to open this model using router:
`/#/5/limit-orders/WETH?tab=history&orderId=0xd2b819`
When there is an `orderId` paramer - open modal

shoom3301 replied:
If it's not so easy to implement now, you can just add TODO for the future

nenadV91 replied:
Added todo :+1:

shoom3301 commented on src/cow-react/modules/limitOrders/hooks/useGetInitialPrice.ts:
```diff
@@ -11,7 +11,8 @@ import { DEFAULT_DECIMALS } from 'custom/constants'
 type PriceResult = number | Error | undefined
 
-const parsePrice = (price: number, currency: Currency) => price * 10 ** (DEFAULT_DECIMALS + getDecimals(currency))
+export const parsePrice = (price: number, currency: Currency) =>
```
Would be better to move the function to `utils` directory

shoom3301 commented on src/cow-react/modules/limitOrders/containers/OrdersWidget/hooks/useLimitOrdersList.ts:
```diff
@@ -1,11 +1,30 @@
+}
+
+export interface ParsedOrder extends Order {
+  executedBuyAmount?: JSBI
```
Why all parameters are optional?

nenadV91 replied:
Mainly because apiAdditionalInfo is optional on order

shoom3301 replied:
But in `useLimitOrdersList()` I see that you always set all parameters except of:
```
surplusFee?: string
partiallyFilled?: boolean
fullyFilled?: boolean
```

shoom3301 commented on src/cow-react/modules/limitOrders/containers/OrdersWidget/hooks/useLimitOrdersList.ts:
```diff
@@ -1,11 +1,30 @@
-import { useMemo } from 'react'
+import { useCallback, useMemo } from 'react'
 import { Order, OrderStatus } from 'state/orders/actions'
+import BigNumber from 'bignumber.js'
```
Can we use `@ethersproject/bignumber` or `JSBI` instead? I think we don't need several implementation of bignumber

nenadV91 replied:
Not sure about JSBI, because it will complain about the decimals and we have decimals in some of these values. Its not ideal that we are mixing these types but I suggest we fix this in the future.

shoom3301 commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+  return JSBI.multiply(JSBI.BigInt(amount), JSBI.BigInt(10 ** decimals))
+}
+
+export function LimitOrdersReceiptModal() {
```
<img width="1231" alt="image" src="https://user-images.githubusercontent.com/7122625/203937281-7259b14f-861f-4459-84c5-0172e1cefe85.png">

Can't scroll to the bottom when the window height is less than 600px

shoom3301 commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/CurrencyField.tsx:
```diff
@@ -0,0 +1,22 @@
+      <styledEl.FieldTitle>{label}</styledEl.FieldTitle>
+
+      <styledEl.FieldBody>
+        <CurrencySelectButton readonlyMode={true} loading={false} currency={token} />
```
P.s. there is `common/pure/CurrencyInputPanel/CurrencyPreview.tsx`, might be helpful

nenadV91 replied:
Tried to use that one but had some issues adjusting it so I just created a new component.

shoom3301 commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/FeeField.tsx:
```diff
@@ -0,0 +1,30 @@
+    quoteSymbol = inputToken.symbol
+  }
+
+  if (!quoteSymbol) {
```
```typescript
const quoteSymbol = inputToken.symbol

if (!quoteSymbol) return null

const formattedExecutedFee = formatSmart(CurrencyAmount.fromRawAmount(inputToken, executedFeeAmount?.toString() || 0))

return (
  <styledEl.Value>
    <span>
      {formattedExecutedFee} {quoteSymbol}
    </span>
 </styledEl.Value>
)
```

shoom3301 commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/index.tsx:
```diff
@@ -0,0 +1,131 @@
+  ORDER_ID = 'Order id TODO',
+}
+
+export function ReceiptModal({
```
Looks very clean and elegantly! Great job!

nenadV91 replied:
Thank you!

shoom3301 commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/PriceField.tsx:
```diff
@@ -0,0 +1,26 @@
+    <styledEl.Value>
+      {price ? (
+        <styledEl.RateValue>
+          1 {order.inputToken.symbol} ={' '}
```
Please, use `common/pure/RateInfo/index.tsx` or add TODO comment for the future

shoom3301 commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/styled.ts:
```diff
@@ -0,0 +1,145 @@
+  background: ${({ theme }) => theme.input.bg1};
+  margin-bottom: 5px;
+  width: 100%;
+  border-radius: ${({ border }) => {
```
You can wrap all fields into one `<div>` and set border-radius: 1rem to the wrapper

shoom3301 commented on src/cow-react/modules/limitOrders/state/limitOrdersReceiptAtom.ts:
```diff
@@ -0,0 +1,18 @@
+  selected: ParsedOrder | null
+}
+
+export const receiptAtom = atom<ReceiptState>({
```
Would be great to replace it by `Router` usage (`?orderId=xxxx`)

nenadV91 replied:
Added todo

shoom3301 commented on src/cow-react/modules/limitOrders/utils/getOrderFilledAmount.ts:
```diff
@@ -0,0 +1,34 @@
+  percentage: BigNumber
+}
+
+export function getOrderFilledAmount(order: Order): FilledAmountResult {
```
```typescript
export function getOrderFilledAmount(order: Order): FilledAmountResult {
  if (!order.apiAdditionalInfo) {
    return { amount: ZERO_BIG_NUMBER, percentage: ZERO_BIG_NUMBER }
  }

  if (order.kind === 'buy') {
    const amount = new BigNumber(order.apiAdditionalInfo.executedBuyAmount)
    const totalAmount = new BigNumber(order.buyAmount.toString())

    return { amount, percentage: amount.div(totalAmount) }
  }

  const amount = new BigNumber(order.apiAdditionalInfo.executedSellAmount).minus(
    order.apiAdditionalInfo?.executedFeeAmount
  )
  const totalAmount = new BigNumber(order.sellAmount.toString())

  return { amount, percentage: amount.div(totalAmount) }
}
```

shoom3301 commented on src/cow-react/modules/limitOrders/utils/getOrderFilledAmount.ts:
```diff
@@ -0,0 +1,34 @@
+import BigNumber from 'bignumber.js'
```
Also would be great to replace by `ethersproject/bignumber` or `JSBI`

shoom3301 reviewed (changes requested):
Awesome job! Code is very clean and understandable.
Added some comments

elena-zh reviewed (changes requested):
Hey @nenadV91 , great job!!
However, I have found several issues in the PR: 
1. It would be great to show titles with full amounts on hover on From, TO, Surplus, Filled, Fee fields
2. It would be nice to add a space to separate percentage and the surplus value
![image](https://user-images.githubusercontent.com/70885163/203981636-8e0f4a55-1ea2-4180-bea0-bc49785428ed.png)

3. Amounts do not match with amount on Explorer for the order details (Filled, From or To columns): 
![not match](https://user-images.githubusercontent.com/70885163/203981736-b28bae32-579f-43bf-b0ea-47798ddfa993.png)
4. There is a typo in the 'Kill' field (double l)
5. There is no dollar estimation in the From/To/Fee fields (they present on the mock-up)
![image](https://user-images.githubusercontent.com/70885163/203982089-48c0d424-df7f-4bb2-85dd-795d8d67be85.png)
6. Order type is missing (Buy or Sell). Besides, there on order is partially fillable, Fill or Kill is still displayed on Order details
![no fill or kill](https://user-images.githubusercontent.com/70885163/203982525-d7971c76-80fc-462b-9cbd-0b8fcde3a75b.png)
7. It would be great to hide a token value when there is no Fee
![no fee](https://user-images.githubusercontent.com/70885163/203982648-c94fd809-f281-4fcb-85ba-1687fa2a1b05.png)
8. From/To fields look weird in Mobile/Tablet views
![tablet](https://user-images.githubusercontent.com/70885163/203982747-9742bdc1-66d9-4dd0-8e54-0e1c91474dce.png)
9. The status is not auto-updated when the modal is opened
![not auto](https://user-images.githubusercontent.com/70885163/203982852-975a9c4c-e355-4b53-9674-67e0ced48262.png)


Thanks!

anxolin commented on package.json:
```diff
@@ -267,6 +267,7 @@
     "jsbi": "^3.1.4",
     "lightweight-charts": "^3.3.0",
     "make-plural": "^7.0.0",
+    "moment": "^2.29.4",
```
Are we sure we want to add this library? 

Last time I check, moment was a bit heavy, and there was others like https://date-fns.org/ more light weight. 

Also, did you review if we have something already, I would swear for the token we showed relative dates

anxolin replied:
indeed. It's important to make sure we don't increase the bundle size, and also we use the libraries we have already

timeago.js is a dependency

https://pub.dev/packages/timeago

we even have a hook https://github.com/cowprotocol/cowswap/blob/d254296019a6841883a2a459a936c3e25f131f53/src/custom/hooks/useTimeAgo.ts#L4

nenadV91 replied:
Replaced moment.js with `date-fns` :+1:

anxolin commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/index.tsx:
```diff
@@ -0,0 +1,130 @@
+  CREATED = 'Created date TODO',
+  EXPIRY = 'Expiry date TODO',
+  ORDER_TYPE = 'Order type TODO',
+  ORDER_ID = 'Order id TODO',
```
We actually have a good initial proposal for the tooltips, you only need to piggyback to explorer

nenadV91 replied:
Will do this in separate PR after

anxolin reviewed (commented):
Looks great! 

I will continue with the review (so expect more comments :P ), but just wanted to send some initial feedback. 

The most important meta-feedback is: This PR is way to big. I'm sure you can organise It to do it in smaller steps. With this big PRs is easy to miss important things.

anxolin commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+  })
+
+  return (
+    <ReceiptModalPure
```
I think we don't call pure components `Pure`, should it be called `ReceiptModal`

nenadV91 replied:
Good catch, updated!

anxolin commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+}
+
+export function LimitOrdersReceiptModal() {
+  const { selected: order } = useAtomValue(receiptAtom)
```
will this component be only for Limit Orders, or for all orders?

nenadV91 replied:
Will update this :+1:

anxolin commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+import { CurrencyAmount, Fraction, Token } from '@uniswap/sdk-core'
+import JSBI from 'jsbi'
+
+function calculatePrice({
```
nit: could you make a type with the params? this way the function definition is smaller, and we have some reusable type we can use.

If you do that, it wouldn't be longer needed, but when the function definition is too long, i prefer to extract the parameters in the function itself (instead of in the definition). The reason again is to make the definition, small and easy to read.

anxolin commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+  inputToken: Token
+  outputToken: Token
+}) {
+  const isZero = (x: JSBI) => JSBI.equal(x, JSBI.BigInt(0))
```
move to utils?

nenadV91 replied:
Done :+1:

anxolin commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+  return new Fraction(adjustedBuy, adjustedSell)
+}
+
+const adjustDecimals = (amount: string | JSBI, decimals: number) => {
```
It's not clear what is this function given the name.

It's a function that calculate the `atoms` given the `decimal` amount? I think we need a better name, and some JS docs. 

Also, we should move to the utils, and make some tests to assert that it does exactly what you expect. 

Lastly, I would be surprised if we don't have a similar util somewhere

nenadV91 replied:
Removed this function since its not needed anymore

anxolin commented on src/cow-react/modules/limitOrders/containers/OrdersWidget/hooks/useLimitOrdersList.ts:
```diff
@@ -15,11 +34,31 @@ export function useLimitOrdersList(): LimitOrdersList {
   const accountLowerCase = account?.toLowerCase()
 
+  const ordersFilter = useCallback((order: Order) => order.owner.toLowerCase() === accountLowerCase, [accountLowerCase])
+  const ordersSorter = (a: Order, b: Order) => Date.parse(b.creationTime) - Date.parse(a.creationTime)
```
I don't think we need to create this function every time, right. It's purely static

should this be in a some utils?

nenadV91 replied:
In my opinion this makes it clean, but no need to move this to utils

anxolin replied:
- Can you elaborate why is cleaner to declare a function inline when it can be a static function? 
- What do you do if you need to sort by creation time somewhere else too? Will you also implement it inline? 
- Unrelated nit: I don't think `a` and `b` are great names for vars.

anxolin commented on src/cow-react/modules/limitOrders/containers/OrdersWidget/hooks/useLimitOrdersList.ts:
```diff
@@ -15,11 +34,31 @@ export function useLimitOrdersList(): LimitOrdersList {
 
+  const ordersFilter = useCallback((order: Order) => order.owner.toLowerCase() === accountLowerCase, [accountLowerCase])
+  const ordersSorter = (a: Order, b: Order) => Date.parse(b.creationTime) - Date.parse(a.creationTime)
+  const ordersParser = (order: Order): ParsedOrder => {
```
I have some questions about this function.

- 1. Is it just to get some augmented order info? 
- 2. If so, why `Parsed`? that makes me very confused. Parsing would be for example to converting from saw raw string into a data object for example. If it's just one type with additional data for the receipt, you could just call it `OrderReceipt`? 
- 3. Are we sure this props shouldn't just be created in the presentation layer instead of in this hook?
- 4. This function looks static, why did we define it inside the hook?

nenadV91 replied:
Hmm this is a tricky one, maybe we can refactor this better

anxolin commented on src/cow-react/modules/limitOrders/containers/OrdersWidget/index.tsx:
```diff
@@ -90,6 +91,7 @@ export function OrdersWidget() {
         showOrderCancelationModal={showOrderCancelationModal}
       />
       {cancelModalProps && <CancellationModal {...cancelModalProps} />}
+      <LimitOrdersReceiptModal />
```
No action needed here, but just to keep in mind. It's likely that we will want to add in the near future a setting to show the orders in the Swap too. 

When this happens, this modal should be moved to a more central place, and it should be called just `OrderReceiptModal`

nenadV91 replied:
Changed the name to `OrderReceiptModal` and will keep this in mind

anxolin commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/index.tsx:
```diff
@@ -0,0 +1,133 @@
+  sellAmount,
+  buyAmount,
+  limitPrice,
+  executionPrice,
```
ubernit, and probably personal preference, but with so many params, i prefer to move the destructuring to the implementation (not in the funciton definition).

I like to have simple function definitions because sometimes you just want to quicly get an idea of the signatures (read the name, param type, and return type) and you don't want to go over too many lines to do that. 

Having it in the body makes it very easy also to see the props used, if you want to review that after checking the signature.

anxolin commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/index.tsx:
```diff
@@ -0,0 +1,133 @@
+
+            <styledEl.Field>
+              <FieldLabel label="Limit price" tooltip={Tooltip.LIMIT_PRICE} />
+              <PriceField order={order} price={limitPrice} />
```
U really like you splitted this into smaller components. Makes it easy to mantain!

nenadV91 replied:
Yep, I've started with one one component but then it grew quickly so I refactored it.

anxolin commented on src/cow-react/modules/limitOrders/pure/ReceiptModal/index.tsx:
```diff
@@ -0,0 +1,133 @@
+}
+
+export function ReceiptModal({
+  isOpen,
+  onDismiss,
```
Would be important to have a cosmos fixture for this?

nenadV91 replied:
Totally, added TODO for this.

anxolin replied:
I think its also the way that we can create all the special cases we want to handle (like partials, limit orders, swaps, etc.) and make sure we style them right

anxolin commented on src/cow-react/modules/limitOrders/utils/getOrderExecutedAmounts.ts:
```diff
@@ -0,0 +1,28 @@
+  }
+
+  return {
+    executedBuyAmount: JSBI.BigInt(order.apiAdditionalInfo.executedBuyAmount),
```
too much repetition of `apiAdditionalInfo`, once you asserted the existence of this, can't you just destrucure it and get `executedBuyAmount`, `executedSellAmount` and `executedFeeAmount`

nenadV91 replied:
Good point, I was in a hurry to modify code from Explorer and to just get this to work but this should be cleaner. Updated.

anxolin commented on src/cow-react/modules/limitOrders/utils/getOrderExecutedAmounts.ts:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param order The order
+ */
+export function getOrderExecutedAmounts(order: Order): {
```
why this is in `limitOrders/utils`, is it really a util only for limit orders?

anxolin commented on src/cow-react/modules/limitOrders/utils/getOrderExecutedAmounts.ts:
```diff
@@ -0,0 +1,28 @@
+
+  return {
+    executedBuyAmount: JSBI.BigInt(order.apiAdditionalInfo.executedBuyAmount),
+    executedSellAmount: JSBI.subtract(
```
Is this correct? `executedSellAmount = executedSellAmount - executedFeeAmount`

Just kidding, but let me explain my point matematically :P 

Given any `executedFeeAmount > 0`

`executedSellAmount = executedSellAmount - executedFeeAmount`

`executedSellAmount - executedSellAmount = executedFeeAmount`

`0 = executedFeeAmount` --> 🔥 false, because `executedFeeAmount > 0`

anxolin replied:
So, should this be `executedSellAmountAfterFee` or sth?

anxolin commented on src/cow-react/modules/limitOrders/utils/getOrderFilledAmount.ts:
```diff
@@ -0,0 +1,34 @@
+  percentage: BigNumber
+}
+
+export function getOrderFilledAmount(order: Order): FilledAmountResult {
+  let executedAmount, totalAmount
```
I wonder if these functions are a direct replication of what we do in the explorer. 

Did you check there the implementation.

Would it make sense to make these things in the SDK, or this adds too much overhead?

shoom3301 commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+
+  const adjustedBuy = adjustDecimals(buyAmount, inputToken.decimals)
+  const adjustedSell = adjustDecimals(sellAmount, outputToken.decimals)
+  return new Fraction(adjustedBuy, adjustedSell)
```
Let's use `import { Price } from '@uniswap/sdk-core'`

nenadV91 replied:
Nice, will use this :+1:

nenadV91 commented:
@elena-zh 
1. Fixed
2. Fixed
3. Work in progress
4. Fixed
5. I think we can skip this for now and add it later
6. Fixed
7. Fixed
8. Fixed
9. This is something we can do in the future

elena-zh commented:
Heya @nenadV91 , great job!
I have created #1533 and #1534 issues to fix cases 5 and 9. 
All the rest fixed cases LGTM!
Just some minor notes (or let me know if I need to create separate issues for them)
1. I would add an order type to an order receipt (Market/Limit/Liquidity). For now, it will be the only one possible way to understand an order type as we don't show a label on the table)
![order type](https://user-images.githubusercontent.com/70885163/204524337-53c5b205-d6f9-4437-af6e-85178d8a3c40.png)
2. It would be nice to enhance the scrollbar on a reduced screen 
![scrollbar](https://user-images.githubusercontent.com/70885163/204524419-2b613a96-67d0-46ee-ac7c-5f03df689c34.png)

And I'm waiting for your fix of the 3rd issue.

Thanks!

nenadV91 commented on src/cow-react/modules/limitOrders/containers/LimitOrdersReceiptModal/index.tsx:
```diff
@@ -0,0 +1,80 @@
+  sellAmount: JSBI | undefined
+  inputToken: Token
+  outputToken: Token
+}) {
```
Will do :+1: 
Moved this to separate util function called `calculatePrice`

nenadV91 commented:
@elena-zh 
In this PR https://github.com/cowprotocol/cowswap/pull/1544
- I've added order class to order type field
- Fixes the order cancel button
- Regarding the scroll bar issue, I don't think that is an issue because for such small screen you should see a native mobile scrollbar like this, and for that you can click on `Toggle device toolbar` in dev-tools
![Screenshot from 2022-11-29 14-42-28](https://user-images.githubusercontent.com/34926005/204583099-b4cd85d8-e0ff-40dc-b8f0-5436184a3980.png)

elena-zh commented:
Hey @nenadV91 , thanks for the explanation. 
I have approved #1544 PR

shoom3301 reviewed (approved):
Approve, but all open issues must be addressed after the release

elena-zh reviewed (approved):
Hey @nenadV91 , great job!
I have found a low-prio issue #1561 , but it should be addressed later.
Thanks!
